File: ruby/02.rb
#!/usr/bin/env ruby

input = File.open(File.expand_path("../inputs/input02.txt", __dir__), "r")
input = input.each_line.map do |line|
    /(?<lo>\d+)-(?<hi>\d+) (?<letter>[a-z]+): (?<password>[a-z]+)/ =~ line.chomp

    {:lo => lo.to_i,
     :hi => hi.to_i,
     :letter => letter,
     :password => password}
end

def is_valid_part1(password, letter, lo, hi)
    num = password.count(letter)
    num >= lo && num <= hi
end

def is_valid_part2(password, letter, lo, hi)
    first = password[lo-1] == letter ? 1 : 0
    second = password[hi-1] == letter ? 1 : 0
    first + second == 1
end

# part 1
puts input.select{|inp| is_valid_part1(inp[:password], inp[:letter], inp[:lo], inp[:hi])}.length

# part 2
puts input.select{|inp| is_valid_part2(inp[:password], inp[:letter], inp[:lo], inp[:hi])}.length

File: elixir/02.exs
defmodule Day02 do
    @moduledoc false

    defp read_input do
        "../inputs/input02.txt"
        |> Path.expand
        |> File.read!
        |> String.split("\n")
    end

    @doc """
    @param {String} - line
    @returns {Map}
    """
    defp regex_parse_line(line) do
        captures = Regex.named_captures(
            ~r/(?<lo>\d+)-(?<hi>\d+) (?<letter>[a-z]+): (?<password>[a-z]+)/,
            String.trim(line)
        )
        %{
            :hi => captures["hi"] |> String.to_integer,
            :lo => captures["lo"] |> String.to_integer,
            :letter => captures["letter"],
            :password => captures["password"]
        }
    end

    @doc """
    Valid if letter appears in password between lo and hi times
    @param {Map} - params
    @returns {Boolean}
    """
    defp is_valid_part1(params) do
        %{:password => password, :letter => letter, :lo => lo, :hi => hi} = params

        count = password
        |> String.graphemes
        |> Enum.count(fn c -> c == letter end)

        count >= lo && count <= hi
    end

    @doc """
    Valid if lo'th XOR hi'th index of password is letter
    @param {Map} - params
    @returns {Boolean}
    """
    defp is_valid_part2(params) do
        %{:password => password, :letter => letter, :lo => lo, :hi => hi} = params


        first = letter == String.at(password, lo-1)
        second = letter ==  String.at(password, hi-1)
        # xor
        (first or second) and not (first and second)
    end

    def solve do
        input = read_input()
        |> Enum.map(&regex_parse_line/1)

        # part 1
        input
        |> Enum.filter(&is_valid_part1/1)
        |> Enum.count
        |> IO.puts

        # part 2
        input
        |> Enum.filter(&is_valid_part2/1)
        |> Enum.count
        |> IO.puts
    end
end
